geonode/api/tests.py

```diff
@@ -226,3 +236,222 @@ def test_date_filter(self):
+                'style_legend_url'
+            ]
+        for f in field_list:
+            self.assertTrue(f in obj)
+
+        if check_ogc_backend(geoserver.BACKEND_PACKAGE):
+            self.assertEqual(obj['type'], 'sld')
+        elif check_ogc_backend(qgis_server.BACKEND_PACKAGE):
+            self.assertEqual(obj['type'], 'qml')
+
+        # Check style detail
+        detail_url = obj['resource_uri']
+        resp = self.api_client.get(detail_url)
+        self.assertValidJSONResponse(resp)
+        obj = self.deserialize(resp)
+
+        # should include body field
+        self.assertTrue('body' in obj and obj['body'])
+
+    def test_add_delete_styles(self):
+        """Style API Add/Delete interaction."""
+        # Check styles count
+        style_list_url = reverse(
+            'api_dispatch_list',
+            kwargs={
+                'api_name': 'api',
+                'resource_name': 'styles'
+            }
+        )
+        filter_url = style_list_url + '?layer__name=' + self.layer.name
+        resp = self.api_client.get(filter_url)
+        self.assertValidJSONResponse(resp)
+        objects = self.deserialize(resp)['objects']
+
+        self.assertEqual(len(objects), 1)
+
+        # Fetch default style
+        layer_detail_url = reverse(
+            'api_dispatch_detail',
+            kwargs={
+                'api_name': 'api',
+                'resource_name': 'layers',
+                'id': self.layer.id
+            }
+        )
+        resp = self.api_client.get(layer_detail_url)
+        self.assertValidJSONResponse(resp)
+        obj = self.deserialize(resp)
+        # Take default style url from Layer detail info
+        default_style_url = obj['default_style']
+        resp = self.api_client.get(default_style_url)
+        self.assertValidJSONResponse(resp)
+        obj = self.deserialize(resp)
+        style_body = obj['body']
+
+        style_stream = StringIO(style_body)
+        # Add virtual filename
+        style_stream.name = 'style.qml'
+        data = {
+            'layer__id': self.layer.id,
+            'name': 'new_style',
+            'title': 'New Style',
+            'style': style_stream
+        }
+        # Use default client to request
+        resp = self.client.post(style_list_url, data=data)
+
+        # Should not be able to add style without authentication
+        self.assertEqual(resp.status_code, 403)
+
+        # Login using anonymous user
+        self.client.login(username='AnonymousUser')
+        style_stream.seek(0)
+        resp = self.client.post(style_list_url, data=data)
+        # Should not be able to add style without correct permission
+        self.assertEqual(resp.status_code, 403)
+        self.client.logout()
+
+        # Use admin credentials
+        self.client.login(username='admin', password='admin')
+        style_stream.seek(0)
+        resp = self.client.post(style_list_url, data=data)
+        self.assertEqual(resp.status_code, 201)
```

Comment:
You can only send post/delete request if you already logged in You need to maintain your session.